src/Business Foundation/App/NoSeries/src/Single/NoSeriesSequenceImpl.Codeunit.al

```diff
@@ -0,0 +1,266 @@
+    end;
+
+    local procedure GetNextNoInternal(var NoSeriesLine: Record "No. Series Line"; ModifySeries: Boolean; UsageDate: Date; HideErrorsAndWarnings: Boolean): Code[20]
+    var
+        NoSeriesLine2: Record "No. Series Line";
+        NoSeriesStatelessImpl: Codeunit "No. Series - Stateless Impl.";
+        NewNo: BigInteger;
+    begin
+        if UsageDate = 0D then
+            UsageDate := WorkDate();
+
+        if not TryGetNextSequenceNo(NoSeriesLine, ModifySeries, NewNo) then begin
+            if not NumberSequence.Exists(NoSeriesLine."Sequence Name") then
+                CreateNewSequence(NoSeriesLine);
+            TryGetNextSequenceNo(NoSeriesLine, ModifySeries, NewNo);
+        end;
+
+        NoSeriesLine2 := NoSeriesLine;
+        NoSeriesLine2."Last No. Used" := GetFormattedNo(NoSeriesLine, NewNo);
+        NoSeriesLine2.Validate(Open);
+
+        if not NoSeriesStatelessImpl.EnsureLastNoUsedIsWithinValidRange(NoSeriesLine2, HideErrorsAndWarnings) then
+            exit('');
+
+        if ModifySeries and ((NoSeriesLine."Last Date Used" <> UsageDate) or (NoSeriesLine.Open <> NoSeriesLine2.Open)) then begin // Only modify the series if either the date or the open status has changed. Otherwise avoid locking the record.
+            NoSeriesLine."Last Date Used" := UsageDate;
+            NoSeriesLine.Open := NoSeriesLine2.Open;
+            NoSeriesLine.Modify(true);
+        end;
+
+        exit(NoSeriesLine2."Last No. Used");
+    end;
+
+    [TryFunction]
+    local procedure TryGetNextSequenceNo(var NoSeriesLine: Record "No. Series Line"; ModifySeries: Boolean; var NewNo: BigInteger)
+    begin
+        if ModifySeries then begin
+            NewNo := NumberSequence.Next(NoSeriesLine."Sequence Name");
+            if NewNo < NoSeriesLine."Starting Sequence No." then
+                NewNo := NumberSequence.Next(NoSeriesLine."Sequence Name");
+        end else begin
+            NewNo := NumberSequence.Current(NoSeriesLine."Sequence Name");
+            NewNo += NoSeriesLine."Increment-by No.";
+        end;
+    end;
+
+    local procedure CreateNewSequence(var NoSeriesLine: Record "No. Series Line")
+    begin
+        CreateNewSequence(NoSeriesLine, NoSeriesLine."Starting Sequence No.");
+    end;
+
+    local procedure CreateNewSequence(var NoSeriesLine: Record "No. Series Line"; StartingSequenceNo: BigInteger)
+    begin
+        if NoSeriesLine."Sequence Name" = '' then
+            NoSeriesLine."Sequence Name" := Format(CreateGuid(), 0, 4);
+
+        NoSeriesLine."Starting Sequence No." := StartingSequenceNo;
+        NumberSequence.Insert(NoSeriesLine."Sequence Name", StartingSequenceNo - NoSeriesLine."Increment-by No.", NoSeriesLine."Increment-by No.");
+        if NumberSequence.Next(NoSeriesLine."Sequence Name") = 0 then; // Get a number to make sure LastNoUsed is set correctly
+    end;
+
+    local procedure GetFormattedNo(NoSeriesLine: Record "No. Series Line"; Number: BigInteger): Code[20]
```

Comment:
This looks like a duplicate from one of the other codeunits. Any way to centralize them instead?

Reply:
It's in one other place and that one is being obsoleted. Furthermore this one is local and we would like to keep it as such since we want to get rid of this function when possible.